Validation/HLTrigger/plugins/HLTGenValSource.cc

```diff
@@ -0,0 +1,388 @@
+  inputCollections.add<edm::InputTag>("ak4GenJets", edm::InputTag("ak4GenJets"));
+  inputCollections.add<edm::InputTag>("ak8GenJets", edm::InputTag("ak8GenJets"));
+  inputCollections.add<edm::InputTag>("TrigEvent", edm::InputTag("hltTriggerSummaryAOD"));
+  desc.add<edm::ParameterSetDescription>("inputCollections", inputCollections);
+
+  // hist descriptors, which are a vector of PSets
+
+  // defining single histConfig
+  // this is generally without default, but a default set of histConfigs is specified below
+  edm::ParameterSetDescription histConfig;
+  histConfig.add<std::string>("vsVar");
+  histConfig.add<std::vector<double>>("binLowEdges");
+  histConfig.addVPSet("rangeCuts", VarRangeCut<HLTGenValObject>::makePSetDescription(), std::vector<edm::ParameterSet>());
+
+  // default set of histConfigs
+  std::vector<edm::ParameterSet> histConfigDefaults;
+
+  edm::ParameterSet histConfigDefault0;
+  histConfigDefault0.addParameter<std::string>("vsVar", "pt");
+  std::vector<double> defaultPtBinning{ 0,5,10,12.5,15,17.5,20,22.5,25,30,35,40,45,50,60,80,100,150,200,250,300,350,400 };
+  histConfigDefault0.addParameter<std::vector<double>>("binLowEdges", defaultPtBinning );
+  histConfigDefaults.push_back(histConfigDefault0);
+
+  edm::ParameterSet histConfigDefault1;
+  histConfigDefault1.addParameter<std::string>("vsVar", "eta");
+  std::vector<double> defaultetaBinning{ -10, -8, -6, -5, -4, -3, -2, -1, 0, 1, 2, 3, 4, 5, 6, 8, 10 };
+  histConfigDefault1.addParameter<std::vector<double>>("binLowEdges", defaultetaBinning );
+  histConfigDefaults.push_back(histConfigDefault1);
+
+  desc.addVPSet("histConfigs", histConfig, histConfigDefaults);
+
+  // defining single histConfig2D
+  edm::ParameterSetDescription histConfig2D;
+  histConfig2D.add<std::string>("vsVarX");
+  histConfig2D.add<std::string>("vsVarY");
+  histConfig2D.add<std::vector<double>>("binLowEdgesX");
+  histConfig2D.add<std::vector<double>>("binLowEdgesY");
+
+  // default set of histConfigs
+  std::vector<edm::ParameterSet> histConfigDefaults2D;
+
+  edm::ParameterSet histConfigDefault2D0;
+  histConfigDefault2D0.addParameter<std::string>("vsVarX", "pt");
+  histConfigDefault2D0.addParameter<std::string>("vsVarY", "eta");
+  histConfigDefault2D0.addParameter<std::vector<double>>("binLowEdgesX", defaultPtBinning);
+  histConfigDefault2D0.addParameter<std::vector<double>>("binLowEdgesY", defaultetaBinning);
+  histConfigDefaults2D.push_back(histConfigDefault2D0);
+
+  desc.addVPSet("histConfigs2D", histConfig2D, histConfigDefaults2D);
+
+  descriptions.addDefault(desc);
+}
+
+// this method handles the different object types and collections that can be used for efficiency calculation
+std::vector<HLTGenValObject> HLTGenValSource::getObjectCollection(const edm::Event& iEvent) {
+
+  std::vector<HLTGenValObject> objects; // the vector of objects to be filled
+
+  // handle object type
+  std::vector<std::string> implemented_GenParticles = {"ele", "pho", "mu", "tau"};
```

Comment:
invalid use of underscore (2.14)

also this is fine but its sort of crying out for an enum...

Reply:
Hi Sam, a question on this: I understand how to define an enum which includes the defined GenParticles. What I am not sure about is how then to compare that to the `objType_` (which is a string from the config file). As far as I understand I would need to write a small function which handles that, as there is no way to directly compare an enum entry and a string...

Is that what you mean, or is there a clever way to do this that I am not seeing?